fix: rewrite winit loop

pietrosophya opened the pull request:
# Objective

- Simplifies/clarifies the winit loop.
- Fixes #12612.

## Solution

The Winit loop runs following this flow:
* NewEvents
* Any number of other events, that can be 0, including RequestRedraw
* AboutToWait

Bevy also uses the UpdateMode, to define how the next loop has to run. It can be essentially:
* Continuous, using ControlFlow::Wait for windowed apps, and ControlFlow::Poll for windowless apps
* Reactive/ReactiveLowPower, using ControlFlow::WaitUntil with a specific wait delay

The changes are made to follow this pattern, so that 
* NewEvents define if the WaitUntil has been canceled because we received a Winit event.
* AboutToWait:
  * checks if the window has to be redrawn
  * otherwise calls app.update() if the WaitUntil timeout has elapsed
  * updates the ControlFlow accordingly

To make the code more logical:
* AboutToWait checks if any Bevy's RequestRedraw event has been emitted
* create_windows is run every cycle, at the beginning of the loop
* the ActiveState (that could be renamed ActivityState) is updated in AboutToWait, symmetrically for WillSuspend/WillResume
* the AppExit events are checked every loop cycle, to exit the app early

alice-i-cecile commented:
Does this also fix #12106?

pietrosophya commented:
I don't have a Windows machine unfortunately.

I tested these on MacOS:
* the game_menu example (using the Quit button)
* the ecs_guide example (windowless, it exits correctly when the game ends)
* the low_power example (using both CTRL+C from the command line, and the window close button)
* the desk_toy example (right click on the logo)

Maximetinu commented on examples/window/low_power.rs:
```diff
@@ -85,7 +85,9 @@ fn update_winit(
             // when there are no inputs, so you send redraw requests while the animation is playing.
-            event.send(RequestRedraw);
+            // Note that in this example the RequestRedraw winit event will make the app run in the same
+            // was as continuous
```
typo here, was -> way ?

BD103 commented:
I tested this on my M1 Macbook, both by using `ci` and by running random examples. It all seems to work!

alice-i-cecile commented:
@pietrosophya I've added a section on platform-specific testing to the PR description: please feel free to update the checklist as reviewer reports come in :)

pietrosophya commented:
Ah, thank you! I also tested WASM on both Chrome and Firefox, do I have to update the list or is it for reviewers only to check platforms off?

alice-i-cecile commented:
Go ahead and check this off yourself :)

mockersf commented:
wasm/webgl2 support is mostly the same across browsers for the event loop, but wasm/webgpu needs to be tested too

hymm commented:
Tested on windows with most of the window example. Seems to work fine.

kristoff3r commented:
Tested every example relating to windows and a few other ones on Linux with X11, no issues found.

wolf-in-space commented:
Tested the window examples on Linux with Wayland and Sway, with the Wayland feature enabled and disabled: Only issue i noticed was window_transparency not working when the Wayland feature is enabled, but the issue is the same on main / 13.1, was also mentioned in #10929 already. Noticed no issues from this pr.

BD103 commented:
WASM works on Safari for WebGL2 :)

pietrosophya commented:
@BD103 I just updated the branch, a few days ago I fixed [another bug](https://github.com/bevyengine/bevy/pull/12683) on WebGPU that was causing the black screen. It works now.

pietrosophya commented:
Solved latest `main` branch conflicts

mockersf commented:
Android doesn't compile on this PR, https://github.com/Sophya/bevy/pull/1 should fix that

mockersf commented:
on Windows, in CI, with this PR, it takes 6 minutes for the app to close after sending the event `AppExit`. This will likely break any attempt to merge because the job will timeout

mockersf commented:
> Do you see any other possible reason? Do you know which test is taking 6 mins to run?

I didn't check every example, but all the ones I checked took around 6 minutes to exit

Also, doesn't work on Android:


https://github.com/bevyengine/bevy/assets/8672791/30065e3b-42e0-4942-9e45-7378a18ed027
[Device logs.txt](https://github.com/bevyengine/bevy/files/14855187/Device.logs.txt)

pietrosophya commented:
I fixed the android window setup. It's not perfect, I think it's still redundant in the way it creates the window and/or the RawWindowHandle. 

NB: the app doesn't work in android in the same way it doesn't in the main branch, so I think it's unrelated to this PR.

pietrosophya commented:
Can anyone help me debug why Windows/CI takes that much to close the app? I have a Mac only and no Windows license near me...

mockersf commented:
works for me on iOS and Android 👍 

I don't have a Windows either, I usually use Bevy modified CI action to run things on Windows

msvbg commented:
Tested on Windows 10. It seems to keep the Window open until I move my mouse, at which point it immediately exits.

Vrixyz commented:
I investigated a bit the issue, I'm no expert in the winit event loop so take with a grain of salt:
- [Original logs](https://gist.github.com/Vrixyz/04dc76176ee4c168ac6c73358009ac99)
  - I tried removing the `return` but that ends up in multiple Exiting events fired [no return](https://gist.github.com/Vrixyz/f3e1054f62450dc769a9adf335f69eef)
  - I tried moving the whole AppExit check after the match(event) produces [correct logs](https://gist.github.com/Vrixyz/096157122de68aafd8e1e11c834be05e) (and exits without delay on windows)
  
<details><summary>Patch for my suggestion on moving AppExit check</summary>
<p>

```patch
diff --git a/crates/bevy_winit/src/lib.rs b/crates/bevy_winit/src/lib.rs
index 3be139185..161082464 100644
--- a/crates/bevy_winit/src/lib.rs
+++ b/crates/bevy_winit/src/lib.rs
@@ -354,13 +354,6 @@ fn handle_winit_event(
         runner_state.redraw_requested = true;
     }
 
-    if let Some(app_exit_events) = app.world().get_resource::<Events<AppExit>>() {
-        if app_exit_event_reader.read(app_exit_events).last().is_some() {
-            event_loop.exit();
-            return;
-        }
-    }
-
     // create any new windows
     // (even if app did not update, some may have been created by plugin setup)
     create_windows(event_loop, create_window.get_mut(app.world_mut()));
@@ -785,6 +778,13 @@ fn handle_winit_event(
         _ => (),
     }
 
+    if let Some(app_exit_events) = app.world().get_resource::<Events<AppExit>>() {
+        if app_exit_event_reader.read(app_exit_events).last().is_some() {
+            event_loop.exit();
+            return;
+        }
+    }
+
     // We drain events after every received winit event in addition to on app update to ensure
     // the work of pushing events into event queues is spread out over time in case the app becomes
     // dormant for a long stretch.
```

</p>
</details> 
  
  I'm not sure that's the correct way of doing it, but I hope sharing my observations is useful 🤞

pietrosophya commented:
I applied this change, it makes some sense since we check for new AppExit events after a cycle has run. But, generally speaking, the cycle should run every time, at most after the WaitUntil timeout. 

I checked on Windows 7, and I wasn't able to reproduce it (I tested clear_color, low_power, multiple_windows, game_menu). How do I run the CI version?

Vrixyz commented:
run `CI_TESTING_CONFIG=.github/example-run/alien_cake_addict.ron cargo run --example load_gltf --features "bevy_ci_testing"` and check how fast it exits without any input

pietrosophya commented:
It closed immediately after 300 frames...

pietrosophya commented:
I think I fixed this: basically, Windows takes some time to create a window, and the loop is put in Poll mode. Then, it was requesting a redraw all the time and never running any App update.

BD103 commented:
Does this introduce any breaking changes that would prevent it from being added to 0.13.3?

pietrosophya commented:
No, it just changes the event loop internals without touching any API.

Il mar 9 apr 2024, 02:59 BD103 ***@***.***> ha scritto:

> Does this introduce any breaking changes that would prevent it from being
> added to 0.13.3?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bevyengine/bevy/pull/12669#issuecomment-2043950612>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZKFRRGJBRLXV5AIPSSHVB3Y4M4PTAVCNFSM6AAAAABFEYZUZOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBTHE2TANRRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

extrawurst commented:
Would love to see this in a patch release

extrawurst commented:
A patch RC with this would be great!

BD103 commented:
> I'm going to want to see this in a release candidate before releasing it in a patch release. It's a really valuable high-impact bug fix, and worth backporting, but too risky to ship without wider testing. I do think this is a candidate for backporting though.

Ok, marking this as 0.14.0 for now so it does not accidentally get added. Perhaps it should be released in 0.13.3-rc.0 or 0.14.0-rc.0? I know aevyrie was discussing publishing release candidates to give plugins more time to migrate.

mockersf commented:
could someone test resuming on Android and iOS?
run the example, move the camera, switch the app to background. music should stop. reopen the app. music should continue where it stopped, camera should be at the same position and still movable

adals commented:
I just want to help 

this is my testing on ios simulator 

https://github.com/bevyengine/bevy/assets/169691/53ccfef3-b7e0-4e65-b132-4ff0dafb64f5

Maximetinu commented:
Hi all

there is something I don't understand about the `EventLoopProxy`

How can the new code of the `low_power.rs` example not to crash, if a resource of the type `EventLoopProxy` was never inserted into the world? 

I was expecting to see a line like `app.init_non_send_resource::<EventLoopProxy>()`. It's not there, but I tried running the example and it of course works

<img width="500" alt="image" src="https://github.com/bevyengine/bevy/assets/16073309/5247ce40-729c-453a-b053-9967c5c6914c">

I must be missing something 🤔

pietrosophya commented:
I made some changes following my updates on a forked 0.13.2: now `should_update` takes care only of window/device events, or the wait timeout elapsed. 

The rendering is instead governed by the `redraw_requested` state when some event causes a new redraw. 

These two (`should_update` and `redraw_requested`) are decoupled since they can be originated from different unrelated causes. I'm even thinking of separating these into two different states.

The second update (or rather a bug fix) is that now the control flow is restored correctly if we set the UpdateMode to Continuous after a Reactive one.

eero-lehtinen commented:
> could someone test suspending and resuming on Android and iOS? run the example, move the camera, switch the app to background. music should stop. reopen the app. music should continue where it stopped, camera should be at the same position and still movable

I tested this right now on android 11 and it works fine except the music doesn't stop at all when closing and resuming. It continues even when I turn the screen off with the power button.

mockersf commented:
Thanks!

> the music doesn't stop at all when closing and resuming. It continues even when I turn the screen off with the power button.

So suspending on Android doesn't work with this PR

pietrosophya commented:
I fixed the suspension in Android, I had to test it on a previous forked version because of [this](https://github.com/bevyengine/bevy/issues/13038). I took another refactoring opportunity and moved some comments to clarify them.

@eero-lehtinen could you try android again?

eero-lehtinen commented:
I works perfectly for me now on Android.

eero-lehtinen commented:
[winit 0.30](https://github.com/rust-windowing/winit/releases/tag/v0.30.0) was released, so maybe we need more rewrites :D

mockersf commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -194,54 +196,51 @@ struct WinitAppRunnerState {
 }
 
+#[derive(PartialEq, Eq, Debug)]
+enum ActivityState {
```
`Activity` here is a name that could be strongly linked to Android, but it isn't. Could you find another name for it?

mockersf commented:
looks OK to me, except for a struct name. Could you change it and solve conflicts?

@Maximetinu I resented your review as the PR change, please re-review if you can

Maximetinu reviewed (approved):
> @Maximetinu I'm requesting a new review as the PR changed, please re-review if you can

Done! It still LGTM, approved ✅ 

There are some more ideas to clean up this part of the code that I think are interesting to consider, but they may be out of scope for this PR, specially if we don't want to introduce breaking changes:

- The difference between UpdateModes `Reactive` and `ReactiveLowPower` is only the kind of events (window and/or device) that should trigger an update, so we could merge them into a single `Reactive`, with 2 booleans to expose this customizability. Maybe even `Continuous` could eventually converge into the same type as well. To keep the same current simplicity to the user, we could expose 2 different constructors for `UpdateMode`: `UpdateMode::reactive()` and `UpdateMode::reactive_low_power()` in the same way we have `WinitSettings::desktop_app()` and `WinitSettings::game()`. The reason for this is that the user may not want to update due to window or device events at all, so the update is controlled manually in some other way, like from an external host app by a request redraw user event through the event loop proxy.

- The fact that the `UserEvent` of the Event Loop Proxy is named `RequestRedraw` makes it easy to confuse with the `WindowEvent::RedrawRequested`. I often mix that one up with `Event::UserEvent(RequestRedraw)`. In fact, since the `RequestRedraw` event can now be sent through the Event Loop Proxy, should it still be an ECS event at all? From the user POV, we now have 2 different ways to request a redraw: through the proxy or through the ECS command, and it's unclear to me if the 2 are exactly the same or not. But I don't have an strong opinion on how to solve this or make this clearer tbh.

pietrosophya commented:
@mockersf I resolved conflicts and changed the name. I'll create another PR after this is merged to make this new `UpdateState` and the ApplicationLifecycle converge since they mean the same thing.

alice-i-cecile reviewed (approved):
I'm happy with the improvements to code quality and bug fixes here, and think this is adequately tested. I'm going to leave an approval and get this merged no later than next Monday to ensure that it gets into the hands of folks to test.

alice-i-cecile commented:
> > error[E0277]: the trait bound `Result<RawHandleWrapper, HandleError>: bevy_ecs::bundle::Bundle` is not satisfied

https://github.com/bevyengine/bevy/actions/runs/8929016345/job/24525847476

@pietrosophya once that's fixed ping me and I'll retry the merge.

pietrosophya commented:
@alice-i-cecile argh, sorry about that, it's fixed now.
